src/components/ble/weather/WeatherService.cpp

```diff
@@ -401,120 +715,74 @@ namespace Pinetime {
       return 0;
     }
 
-    std::unique_ptr<WeatherData::Clouds>& WeatherService::GetCurrentClouds() {
+    std::unique_ptr<WeatherData::TimelineHeader>& WeatherService::GetCurrentEvent(WeatherData::eventtype eventType) {
       uint64_t currentTimestamp = GetCurrentUnixTimestamp();
+      auto* best = this->nullHeader;
       for (auto&& header : this->timeline) {
-        if (header->eventType == WeatherData::eventtype::Clouds && currentTimestamp >= header->timestamp &&
-            IsEventStillValid(header, currentTimestamp)) {
-          return reinterpret_cast<std::unique_ptr<WeatherData::Clouds>&>(header);
+        if (header->eventType == eventType && currentTimestamp >= header->timestamp && IsEventStillValid(header, currentTimestamp)) {
+          if ((*best)->timestamp == 0 || header->expires < (*best)->expires ||
+              (header->expires == (*best)->expires && header->timestamp > (*best)->timestamp)) {
+            best = &header;
+          }
         }
       }
 
-      return reinterpret_cast<std::unique_ptr<WeatherData::Clouds>&>(*this->nullHeader);
+      return *best;
     }
 
-    std::unique_ptr<WeatherData::Obscuration>& WeatherService::GetCurrentObscuration() {
+    std::unique_ptr<WeatherData::TimelineHeader>& WeatherService::GetEventByID(WeatherData::eventtype eventType, uint16_t id) {
       uint64_t currentTimestamp = GetCurrentUnixTimestamp();
       for (auto&& header : this->timeline) {
-        if (header->eventType == WeatherData::eventtype::Obscuration && currentTimestamp >= header->timestamp &&
-            IsEventStillValid(header, currentTimestamp)) {
-          return reinterpret_cast<std::unique_ptr<WeatherData::Obscuration>&>(header);
+        if (header->eventType == eventType && IsEventStillValid(header, currentTimestamp) && header->eventID == id) {
+          return header;
         }
       }
 
-      return reinterpret_cast<std::unique_ptr<WeatherData::Obscuration>&>(*this->nullHeader);
+      return *this->nullHeader;
     }
 
-    std::unique_ptr<WeatherData::Precipitation>& WeatherService::GetCurrentPrecipitation() {
-      uint64_t currentTimestamp = GetCurrentUnixTimestamp();
-      for (auto&& header : this->timeline) {
-        if (header->eventType == WeatherData::eventtype::Precipitation && currentTimestamp >= header->timestamp &&
-            IsEventStillValid(header, currentTimestamp)) {
-          return reinterpret_cast<std::unique_ptr<WeatherData::Precipitation>&>(header);
-        }
-      }
+    std::unique_ptr<WeatherData::Clouds>& WeatherService::GetCurrentClouds() {
+      return reinterpret_cast<std::unique_ptr<WeatherData::Clouds>&>(GetCurrentEvent(WeatherData::eventtype::Clouds));
```

Comment:
The reason I wrote each named getter separately was to make it easy to add custom heuristics based on the data type.

This might not even be necessary, but I thought I'd mention it.

Reply:
I don't think that we'll need custom heuristics for the different data types, but good to know your thinking.